diff --git a/x.c b/x.c
index 48a6676..b3fc58c 100644
--- a/x.c
+++ b/x.c
@@ -658,6 +658,8 @@ setsel(char *str, Time t)
 	XSetSelectionOwner(xw.dpy, XA_PRIMARY, xw.win, t);
 	if (XGetSelectionOwner(xw.dpy, XA_PRIMARY) != xw.win)
 		selclear();
+
+	clipcopy(NULL);
 }
 
 void
@@ -676,6 +678,8 @@ brelease(XEvent *e)
 
 	if (mouseaction(e, 1))
 		return;
+  if (e->xbutton.button == Button2)
+    clippaste(NULL);
 	if (e->xbutton.button == Button1)
 		mousesel(e, 1);
 }
